api/src/main/java/com/github/retrooper/packetevents/event/EventManager.java

```diff
@@ -96,28 +116,55 @@ public PacketListenerCommon registerListener(PacketListenerCommon listener) {
      */
     public PacketListenerCommon[] registerListeners(PacketListenerCommon... listeners) {
         for (PacketListenerCommon listener : listeners) {
-            registerListener(listener);
+            this.registerListenerNoRecalculation(listener);
         }
+        this.recalculateListeners();
         return listeners;
     }
 
     public void unregisterListener(PacketListenerCommon listener) {
-        Set<PacketListenerCommon> listenerSet = listenersMap.get(listener.getPriority());
-        if (listenerSet == null) return;
-        listenerSet.remove(listener);
+        if (this.unregisterListenerNoRecalculation(listener)) this.recalculateListeners();
     }
 
     public void unregisterListeners(PacketListenerCommon... listeners) {
+        boolean modified = false;
         for (PacketListenerCommon listener : listeners) {
-            unregisterListener(listener);
+            modified |= this.unregisterListenerNoRecalculation(listener);//OR - at least one of these needs to be true to return true, meaning the boolean will permanently stay true if once set so
         }
+        if (modified) this.recalculateListeners();
     }
 
 
     /**
      * Unregister all dynamic packet event listeners.
      */
     public void unregisterAllListeners() {
-        listenersMap.clear();
+        this.listenersMap.clear();
+        this.listeners = new PacketListenerCommon[0];
```

Comment:
Shouldnt this call the recalc method to prevent race conditions?

Reply:
You could also just completely replace this implementation:
```java
this.unregisterListeners(this.listeners);
```

Reply:
> Shouldnt this call the recalc method to prevent race conditions?

Oh, right. Both of these are valid solutions. Another possible solution is to add a synchronized(this) block on the update of 'listeners' (both in recalculate and unregisterAll). I think that one would be the fastest, so I'd go with that